File: api/transcribe.js
const https = require('https');

module.exports = async function handler(req, res) {
  if (req.method === 'OPTIONS') {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    return res.status(204).end();
  }

  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' });

  const DG_KEY = process.env.DEEPGRAM_API_KEY || 'ffbf23cb49159c3d5699587c20f1debecb819fc5';

  try {
    const chunks = [];
    for await (const chunk of req) chunks.push(chunk);
    const audio = Buffer.concat(chunks);

    if (audio.length < 100) {
      return res.status(400).json({ error: 'No audio data' });
    }

    console.log('[STT] Received ' + audio.length + ' bytes');

    const result = await new Promise((resolve, reject) => {
      const r = https.request({
        hostname: 'api.deepgram.com',
        path: '/v1/listen?model=nova-2&language=de&smart_format=true',
        method: 'POST',
        headers: {
          'Authorization': 'Token ' + DG_KEY,
          'Content-Type': 'audio/wav',
          'Content-Length': audio.length
        }
      }, resp => {
        let d = '';
        resp.on('data', c => d += c);
        resp.on('end', () => {
          try { resolve(JSON.parse(d)); }
          catch(e) { reject(new Error('Parse error')); }
        });
      });
      r.on('error', reject);
      r.write(audio);
      r.end();
    });

    let t = '';
    const ch = result && result.results && result.results.channels;
    if (ch && ch[0] && ch[0].alternatives && ch[0].alternatives[0]) {
      t = ch[0].alternatives[0].transcript || '';
    }

    console.log('[STT] Result: "' + t + '"');
    res.setHeader('Content-Type', 'application/json');
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.json({ transcript: t });
  } catch (e) {
    console.error('[STT] Error:', e.message);
    res.status(500).json({ error: e.message });
  }
};

File: api/transcribe.py
"""Deepgram STT - Vercel Serverless Function"""
import json
import urllib.request

DG_KEY = 'ffbf23cb49159c3d5699587c20f1debecb819fc5'
DG_URL = 'https://api.deepgram.com/v1/listen?model=nova-2&language=de&smart_format=true'

def handler(request):
    if request.method != 'POST':
        return Response(status=405, body='Method not allowed')

    try:
        body = request.body
        if not body or len(body) < 100:
            return Response(
                status=400,
                body=json.dumps({'error': 'No audio data'}),
                headers={'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'}
            )

        req = urllib.request.Request(
            DG_URL,
            data=body,
            headers={
                'Authorization': 'Token ' + DG_KEY,
                'Content-Type': 'audio/wav'
            }
        )
        with urllib.request.urlopen(req, timeout=15) as resp:
            result = json.loads(resp.read())

        t = ''
        ch = result.get('results', {}).get('channels', [])
        if ch and ch[0].get('alternatives'):
            t = ch[0]['alternatives'][0].get('transcript', '')

        return Response(
            status=200,
            body=json.dumps({'transcript': t}),
            headers={'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'}
        )
    except Exception as e:
        return Response(
            status=500,
            body=json.dumps({'error': str(e)}),
            headers={'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'}
        )


class Response:
    def __init__(self, status=200, body='', headers=None):
        self.status = status
        self.body = body if isinstance(body, str) else json.dumps(body)
        self.headers = headers or {'Content-Type': 'application/json'}
        if 'Content-Length' not in self.headers:
            self.headers['Content-Length'] = str(len(self.body.encode('utf-8') if isinstance(self.body, str) else self.body))
